### PR TITLE
[GLib] Provide and use helper functions to convert GBytes to std::span

### PR DESCRIPTION
**Pull request by aperezdc**
#### bdeb007fd6c1fe75bfc58afdf501ce0ddb5f465f
<pre>
[GLib] Provide and use helper functions to convert GBytes to std::span
<a href="https://bugs.webkit.org/show_bug.cgi?id=282210">https://bugs.webkit.org/show_bug.cgi?id=282210</a>

Reviewed by Geoffrey Garen and Michael Catanzaro.

Add a new span(GBytes*) utility function, and a span(GRefPtr&lt;GBytes&gt;&amp;)
overload. These localize the unsafe creation of a std::span&lt;const uint8_t&gt;
that represents the GBytes data in a single location, allowing to remove
WTF_ALLOW_UNSAFE_BUFFER_USAGE_{BEGIN,END} macros which were sprinkled
around. For item types other than uint8_t it is recommended to use
WTF::spanReinterpretCast&lt;T&gt; afterwards.

* Source/WTF/wtf/PlatformGTK.cmake: List new SpanHelpers.h header.
* Source/WTF/wtf/PlatformWPE.cmake: Ditto.
* Source/WTF/wtf/glib/GSpanExtras.h: Added.
(WTF::span): New function to make a std::span out of a GBytes.
* Source/WebKit/NetworkProcess/cache/NetworkCacheDataGLib.cpp:
(WebKit::NetworkCache::Data::span const): Use WTF::span() helper.
(WebKit::NetworkCache::Data::apply const): Use the Data::span() method.
* Source/WebKit/UIProcess/API/glib/WebKitUserContentFilterStore.cpp:
(webkitUserContentFilterStoreSaveBytes): Use WTF::span() helper.
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkit_web_view_load_bytes): Ditto.
* Source/WebKit/UIProcess/API/gtk/DropTargetGtk4.cpp:
(WebKit::DropTarget::accept): Ditto.
* Source/WebKit/UIProcess/Automation/gtk/WebAutomationSessionGtk.cpp:
(WebKit::base64EncodedPNGData): Ditto.
* Source/WebKit/UIProcess/Inspector/glib/RemoteInspectorHTTPServer.cpp:

Canonical link: <a href="https://commits.webkit.org/285855@main">https://commits.webkit.org/285855@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/8a7565e5eb50e69226318ab99565fad3e1f0ac4c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74026 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53455 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26837 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/78379 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/25264 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62588 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1240 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/78379 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/25264 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77093 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/48351 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/63675 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/78379 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/45213 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21166 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/23597 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/67163 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/66727 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21513 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79918 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/73284 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1343 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/726 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/66515 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1487 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63689 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/65790 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16298 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/9701 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/7874 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/95065 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1307 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/4088 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20886 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/1336 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/1324 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1343 "Built successfully") | | |

### CODE REVIEW

File: Source/WTF/wtf/glib/SpanHelpers.h
Comment (geoffreygaren) [inline Source/WTF/wtf/glib/SpanHelpers.h]: We've been calling these helper functions just "span".

File: Source/WTF/wtf/glib/SpanHelpers.h
Comment (geoffreygaren) [inline Source/WTF/wtf/glib/SpanHelpers.h]: Ditto.

File: Source/WTF/wtf/glib/SpanHelpers.h
Comment (geoffreygaren) [reply Source/WTF/wtf/glib/SpanHelpers.h]: Since the type returned by g_bytes_get_data is a uint8, I'd suggest making this a plain (non-template) function that returns uint8, and using span_reinterpret_cast anywhere you need something different. That way, we establish a separation of concerns between documenting bounds and performing type conversion.

File: Source/WTF/wtf/PlatformWPE.cmake
Comment (geoffreygaren) [inline Source/WTF/wtf/PlatformWPE.cmake]: Can we call this GBytesSPI.h or GBytesExtras.h? We're adding a lot of these helpers and they'll conflict if we give them generic names.

**Review by geoffreygaren — CHANGES_REQUESTED** [review]

File: Source/WTF/wtf/glib/SpanHelpers.h
Comment (aperezdc) [reply Source/WTF/wtf/glib/SpanHelpers.h]: Thanks, I will update taking both suggestions into account 👍🏼

File: Source/WTF/wtf/PlatformWPE.cmake
Comment (aperezdc) [reply Source/WTF/wtf/PlatformWPE.cmake]: I am planning to add more conversion helpers in the future, some of them not related to `GBytes`. How about `GSpanConversion.h` or `GSpanExtras.h`?

**Review by geoffreygaren — APPROVED** [review]
r=me

File: Source/WTF/wtf/PlatformWPE.cmake
Comment (aperezdc) [reply Source/WTF/wtf/PlatformWPE.cmake]: I went with `GSpanExtras.h` in the end.

**Comment by aperezdc** [comment]
> r=me

Thanks, I appreciate it! I'll try and get a second review from a GTK/WPE port developer before landing.

File: Source/WTF/wtf/glib/GSpanExtras.h
Comment (mcatanzaro) [inline Source/WTF/wtf/glib/GSpanExtras.h]: I think one of our unwritten code style guidelines is to not write `std::` for stuff inherited from C:

```suggestion
    size_t size = 0;
```

**Review by mcatanzaro — APPROVED** [review]
Looks good.

**Comment by mcatanzaro** [comment]
The mac-safer-cpp EWS failure is fixed by https://commits.webkit.org/285847@main. Please try rebasing.

File: Source/WTF/wtf/glib/GSpanExtras.h
Comment (mcatanzaro) [inline Source/WTF/wtf/glib/GSpanExtras.h:37]: Hm, in retrospect, I probably would have not added this overload. All it does is save you from writing `.get()` which doesn't seem like a great reason to have twice as many functions here.

Comment (mcatanzaro) [reply Source/WTF/wtf/glib/GSpanExtras.h:37]: I guess this is a pretty minor complaint, but it does add up when adding additional helper functions here, e.g. for https://bugs.webkit.org/show_bug.cgi?id=282239 I currently have:

```
diff --git a/Source/WTF/wtf/glib/GSpanExtras.h b/Source/WTF/wtf/glib/GSpanExtras.h
index 8c7c9d57e23d..623eaeba926c 100644
--- a/Source/WTF/wtf/glib/GSpanExtras.h
+++ b/Source/WTF/wtf/glib/GSpanExtras.h
@@ -36,6 +36,28 @@ inline std::span<const uint8_t> span(const GRefPtr<GBytes>& bytes)
     return span(bytes.get());
 }
 
+inline std::span<const uint8_t> span(GVariant* variant)
+{
+    const auto* ptr = static_cast<const uint8_t*>(g_variant_get_data(variant));
+    size_t size = g_variant_get_size(variant);
+    return unsafeForgeSpan(const uint8_t>(ptr, size));
+}
+
+inline std::span<const uint8_t> span(const GRefPtr<GVariant>& variant)
+{
+    return span(variant.get());
+}
+
+inline std::span<const uint8_t> span(GByteArray* array)
+{
+    return unsafeForgeSpan<const uint8_t>(array->data, array->len);
+}
+
+inline std::span<const uint8_t> span(const GRefPtr<GByteArray>& array)
+{
+    return span(array.get());
+}
+
 } // namespace WTF
 
 using WTF::span;
 ```

Comment (aperezdc) [reply Source/WTF/wtf/glib/GSpanExtras.h:37]: TBH, I would rather write two helper functions one single time, than using `.get()` all over the code base without need. Changing all the:

```cpp
auto data = span(bytes);
```

to:

```cpp
auto data = span(bytes.get());
```

would only add noise, and require more typing each time one of the `span()` helper functions are used.

Comment (aperezdc) [reply Source/WTF/wtf/glib/GSpanExtras.h:37]: The other option is to add an implicit conversion to allow a `GRefPtr<T>` to decay into `T*`, but I am not sure that's a good idea, given that writing a one-line overload while keeping explicit what is happening&mdash;I do not enjoy much the idea of implicit conversions.

Comment (mcatanzaro) [reply Source/WTF/wtf/glib/GSpanExtras.h:37]: Let's not allow implicit conversion. I'll just keep things as-is.

**Pull request merged**
